internal/testing/enginetest/enginetest.go

```diff
@@ -0,0 +1,229 @@
+			{Name: "4", Type: &wasm.FunctionType{}, Body: []byte{wasm.OpcodeEnd}, Module: &wasm.ModuleInstance{}},
+		}
+		tableInit := map[wasm.Index]wasm.Index{0: 0, 1: 4}
+
+		// Imported functions are compiled before the importing module is instantiated.
+		imported, err := e.NewModuleEngine(t.Name(), nil, importedFunctions, nil, nil)
+		require.NoError(t, err)
+		defer imported.Close()
+
+		// Instantiate the importing module, which is whose table is initialized.
+		importing, err := e.NewModuleEngine(t.Name(), importedFunctions, moduleFunctions, table, tableInit)
+		require.NoError(t, err)
+		defer importing.Close()
+
+		// A moduleEngine's compiled function slice includes its imports, so the offsets are absolute.
+		require.Equal(t, table.Table, initTable(importing, table.Min, tableInit))
+	})
+}
+
+func RunTestModuleEngine_Call_HostFn(t *testing.T, newEngine func() wasm.Engine) {
+	memory := &wasm.MemoryInstance{}
+	var ctxMemory publicwasm.Memory
+	hostFn := reflect.ValueOf(func(ctx publicwasm.Module, v uint64) uint64 {
+		ctxMemory = ctx.Memory()
+		return v
+	})
+
+	e := newEngine()
+	module := &wasm.ModuleInstance{Memory: memory}
+	modCtx := wasm.NewModuleContext(context.Background(), wasm.NewStore(e, wasm.Features20191205), module, nil)
+
+	f := &wasm.FunctionInstance{
+		GoFunc: &hostFn,
+		Kind:   wasm.FunctionKindGoModule,
+		Type: &wasm.FunctionType{
+			Params:  []wasm.ValueType{wasm.ValueTypeI64},
+			Results: []wasm.ValueType{wasm.ValueTypeI64},
+		},
+		Module: module,
+		Index:  0,
+	}
+	module.Types = []*wasm.TypeInstance{{Type: f.Type}}
+	module.Functions = []*wasm.FunctionInstance{f}
+
+	me, err := e.NewModuleEngine(t.Name(), nil, module.Functions, nil, nil)
+	require.NoError(t, err)
+
+	t.Run("defaults to module memory when call stack empty", func(t *testing.T) {
+		// When calling a host func directly, there may be no stack. This ensures the module's memory is used.
+		results, err := me.Call(modCtx, f, 3)
+		require.NoError(t, err)
+		require.Equal(t, uint64(3), results[0])
+		require.Same(t, memory, ctxMemory)
+	})
+
+	t.Run("errs when not enough parameters", func(t *testing.T) {
+		_, err := me.Call(modCtx, f)
+		require.EqualError(t, err, "expected 1 params, but passed 0")
+	})
+
+	t.Run("errs when too many parameters", func(t *testing.T) {
+		_, err := me.Call(modCtx, f, 1, 2)
+		require.EqualError(t, err, "expected 1 params, but passed 2")
+	})
+}
+
+// addFunction assigns and adds a function to the module.
+func addFunction(module *wasm.ModuleInstance, fn *wasm.FunctionInstance) {
+	module.Functions = append(module.Functions, fn)
+	// This link is essential for all engines. For example, functions call other functions defined in the same module.
+	fn.Module = module
+}
+
+// newModuleContext creates an internalwasm.ModuleContext for unit tests.
+//
+// Note: This sets fields that are not needed in the interpreter, but are required by code compiled by JIT. If a new
```

Comment:
@mathetake took me a while to re-remember this, so hope it isn't too much to leave this comment in